### PR TITLE
[VCMML-670] Add ability to coarse-grain physics diagnostics

### PR DESCRIPTION
**Pull request by spencerkclark**
This PR adds code to enable coarse-graining of physics diagnostics on either model levels or pressure levels.  It is tested in [this notebook](https://github.com/VulcanClimateModeling/explore/blob/master/spencerc/2020-11-24-verify-physics-diagnostic-coarsening/2020-11-24-verify-physics-diagnostic-coarsening.ipynb).

### CODE REVIEW

**Comment by spencerkclark** [comment]
[Port online coarse-graining routines to fv3gfs-fortran](https://trello.com/c/WsBtxjuO/329-port-online-coarse-graining-routines-to-fv3gfs-fortran)

**Comment by spencerkclark** [comment]
The concerns expressed in the initial description related to where the coarse-graining routines live have been addressed in #33.  I've cleaned up this PR a little and now think it is ready for review.  [This notebook](https://github.com/VulcanClimateModeling/explore/blob/master/spencerc/2020-05-13/2020-05-13-verify-coarse-physics-diagnostics.ipynb) shows that the coarse-graining works as expected for both 2D and 3D fields.

File: FV3/gfsphysics/GFS_layer/GFS_typedefs.F90 [hunk 1]
```diff
@@ -5067,7 +5067,7 @@ subroutine diag_create (Diag, IM, Model)
 !      allocate (Diag%upd_mf (IM,Model%levs))
 !      allocate (Diag%dwn_mf (IM,Model%levs))
 !      allocate (Diag%det_mf (IM,Model%levs))
-!      allocate (Diag%cldcov (IM,Model%levs))
+     allocate (Diag%cldcov (IM,Model%levs))
```
[the remark this answers is not in the document]
Comment (oliverwm1) [reply FV3/gfsphysics/GFS_layer/GFS_typedefs.F90:5070]: Why the `cldcov` in particular?

And got it--I've noticed those temperature tendencies in the physics. Would be great if we can use them, but agree we'll definitely need to do some testing and probably fixing.

File: FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90 [hunk 2]
```diff
@@ -142,6 +144,7 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
     ExtDiag(idx)%cnvfac = cn_one
     ExtDiag(idx)%time_avg = .TRUE.
     ExtDiag(idx)%intpl_method = 'bilinear'
+    ExtDiag(idx)%coarse_graining_method = 'area_weighted'
```
Comment (oliverwm1) [inline FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90:147]: Hard to tell from GitHub interface, but did you add a coarse_graining_method for every physics diagnostic, or some subset?

File: FV3/gfsphysics/GFS_layer/GFS_typedefs.F90 (hunk 1 above)
[the remark this answers is not in the document]
Comment (spencerkclark) [reply FV3/gfsphysics/GFS_layer/GFS_typedefs.F90:5070]: The `cldcov` in particular is referenced in the `diag_rad_zero` subroutine if `ldiag3d` is set to `.true.`, so the model segfaults if memory is not allocated for it:
https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/c0d11c6a66f76835ae0e9bc791d7cda6d68e504b/FV3/gfsphysics/GFS_layer/GFS_typedefs.F90#L5247-L5249

[the remark this answers is not in the document]
Comment (spencerkclark) [reply FV3/gfsphysics/GFS_layer/GFS_typedefs.F90:5070]: I'll make an issue for the temperature tendency diagnostics; see [here](https://trello-attachments.s3.amazonaws.com/5d374a1cf335c515ed1a6fab/5ddc4dcec5fd582f618665f5/915889d8fdb356b6b41f1ecde425843d/2019-11-26-fixes-for-dt3dt-variables.html) for a summary of what I needed to do in SHiELD.

File: FV3/atmos_model.F90 [hunk 3]
```diff
@@ -895,9 +903,13 @@ subroutine update_atmos_model_state (Atmos)
...
       call atmosphere_nggps_diag(Atmos%Time)
-      call FV3GFS_diag_output(Atmos%Time, IPD_DIag, Atm_block, IPD_Control%nx, IPD_Control%ny, &
+      call FV3GFS_diag_output(Atmos%Time, IPD_DIag, Atm_block, IPD_Data, IPD_Control%nx, IPD_Control%ny, &
                             IPD_Control%levs, 1, 1, 1.d0, time_int, time_intfull,              &
-                            IPD_Control%fhswr, IPD_Control%fhlwr)
+                            IPD_Control%fhswr, IPD_Control%fhlwr, &
+                            Atm(mytile)%coarse_graining%write_coarse_diagnostics, &
+                            IPD_Diag_coarse, &
+                            Atm(mytile)%delp(Atm(mytile)%bd%is:Atm(mytile)%bd%ie, &
+                            Atm(mytile)%bd%js:Atm(mytile)%bd%je,1:Atm(mytile)%npz))
```
[the remark this answers is not in the document]
Comment (spencerkclark) [reply FV3/atmos_model.F90:912]: Agreed -- I'll create local variables.

File: FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90 (hunk 2 above)
Comment (spencerkclark) [reply FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90:147]: I did not. Instead I tried to make sure that every diagnostic we would be interested in for a high resolution simulation would have a method specified.  Some other variables (e.g. [`soilm`](https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/c0d11c6a66f76835ae0e9bc791d7cda6d68e504b/FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90#L900-L911)) are spatially masked in various ways before being written out to files, so they may need more care in coarse-graining.

If a user requests a coarse diagnostic, e.g. `soilm_coarse`, that does not have a specified coarse-graining method, the model will crash with an informative error message.

File: FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90 [hunk 4]
```diff
@@ -2226,7 +2280,12 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
     enddo
    ENDIF
 
-
+    idx = idx + 1
+    ExtDiag(idx)%axes = 3
+    ExtDiag(idx)%name = 'delp_phys'
```
[the remark this answers is not in the document]
Comment (spencerkclark) [reply FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90:2284]: That very well could be it.  Do you know of a place where that happens to `Atm%delp` itself?  I know that happens to `IPD_Data(nb)%Statein%prsl`:
https://github.com/VulcanClimateModeling/fv3gfs-fortran/blob/c0d11c6a66f76835ae0e9bc791d7cda6d68e504b/FV3/atmos_cubed_sphere/driver/fvGFS/atmosphere.F90#L2059-L2066

File: FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90 [hunk 5]
```diff
@@ -1054,6 +1080,8 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
     ExtDiag(idx)%mod_name = 'gfs_phys'
     ExtDiag(idx)%intpl_method = 'bilinear'
     allocate (ExtDiag(idx)%data(nblks))
+        ExtDiag(idx)%coarse_graining_method = 'area_weighted'
+
```
[the remark this answers is not in the document]
Comment (spencerkclark) [reply FV3/gfsphysics/GFS_layer/GFS_diagnostics.F90]: Thanks

File: FV3/gfsphysics/GFS_layer/GFS_typedefs.F90 (hunk 1 above)
[the remark this answers is not in the document]
Comment (spencerkclark) [reply FV3/gfsphysics/GFS_layer/GFS_typedefs.F90:5070]: See #38

**Comment by spencerkclark** [comment]
[This updated notebook](https://github.com/VulcanClimateModeling/explore/blob/master/spencerc/2020-11-24-verify-physics-diagnostic-coarsening/2020-11-24-verify-physics-diagnostic-coarsening.ipynb) shows that this PR has now been extended to support pressure-level coarse-graining of 3D diagnostics.

**Pull request merged**